Issue10 商品購入機能

yamachan32 opened the pull request:
商品購入機能を実装しました。
コードレビューをお願いできればと思います。
よろしくお願いします。

# What
- orderとaddressを管理するフォームオブジェクトパターンの導入
- orderコントローラーの修正
- クレジットカード決済機能の導入
- 商品購入情報に関するビューファイルの修正
- 単体テストコードの実装

# Why
- 商品購入機能を実装するため

# Gyazo
- [必要な情報を適切に入力して「購入」ボタンを押すと、商品の購入ができる動画](https://gyazo.com/c0c30d5e3ae97806e12ada4ca50b93f0)
- [入力に問題がある状態で「購入」ボタンが押された場合、情報は受け入れられず、購入ページでエラーメッセージが表示される動画](https://gyazo.com/cdc78f4d58f1a3407b6aee7768b0bed1)
- [ログイン状態の場合でも、URLを直接入力して自身が出品していない売却済み商品の商品購入ページへ遷移しようとすると、トップページに遷移する動画](https://gyazo.com/994d353d61f1341474410b28d444e549)
- [ログイン状態の場合でも、URLを直接入力して自身が出品した商品の商品購入ページに遷移しようとすると、商品の販売状況に関わらずトップページに遷移する動画](https://gyazo.com/5cf5cd9a003aa24730006aa59f580e03)
- [ログアウト状態の場合は、URLを直接入力して商品購入ページに遷移しようとすると、商品の販売状況に関わらずログインページに遷移する動画]()
- [売却済みの商品は、画像上に「sold out」の文字が表示される動画（商品一覧機能実装時に未実装であった場合）](https://gyazo.com/9c40f4a8474d6fedae431f9af4d1a20b)
- [売却済みの商品は、画像上に「sold out」の文字が表示される動画（商品詳細機能実装時に未実装であった場合）](https://gyazo.com/4df805a9bb2d04b4069c67ef119a235b)
- [ログイン状態の場合でも、売却済みの商品には、「商品の編集」「削除」「購入画面に進む」ボタンが表示されない動画（商品詳細機能実装時に未実装であった場合）](https://gyazo.com/c5d4eb7f8a9dd52c1f61d79d1080b96c)
- [ログイン状態の場合でも、URLを直接入力して自身が出品した売却済み商品の商品情報編集ページへ遷移しようとすると、トップページに遷移する動画（商品情報編集機能実装時に未実装であった場合）](https://gyazo.com/4bc6d668f9cc832213abe85e56cf7bda)
- [テスト結果の画像](https://gyazo.com/786d78977b83f634d8bf05bf458c9c68)

yamachan32 committed:
Issue10 商品購入機能 データベース修正

- oderモデル/addressモデルの作成
- アソシエーション/バリデーションの設定
- マイグレーションの実行

yamachan32 committed:
Issue10 商品購入機能 ルーティング設定

- ordersへのルーティングをitemsにネスト

yamachan32 committed:
Issue10 商品購入機能 モデルの作成

- order_addressモデルの作成
- orderモデルの作成
- addressモデルの作成
- 各モデルにアソシエーションとバリデーションを設定

yamachan32 committed:
Issue10 商品購入機能 コントローラーの作成

- ordersコントローラーの作成
- itemsコントローラーに売却済みitemの編集をリダイレクトする記述の追加

yamachan32 committed:
Issue10 商品購入機能 ビューファイルの作成

- orders/indexビューの修正
- item index/showに売却済みアイテムの表示制御を追加

yamachan32 committed:
Issue10 商品購入機能 Gemの追加

- 'payjp'の追加

yamachan32 committed:
Issue10 商品購入機能 クレジットカードトークン取得機能の追加

- webpackerの追加/環境変数の読み込み
- card.jsの作成
- card.jsの読み込み設定
- card.jsにトークン取得の記述を追加

yamachan32 committed:
Issue10 商品購入機能 不要ファイルの削除

- 誤って作成していた不要ファイルを削除

yamachan32 committed:
Issue10 商品購入機能 単体テストコードの追加

- order_addressモデルのFactoryBotを設定
- order_addressモデルのテストコードを作成

yamachan32 committed:
Issue10 商品購入機能 robocopの実行

AsukaToyoshima commented on app/javascript/card.js:
```diff
@@ -0,0 +1,31 @@
+      exp_month: formData.get("order_address[exp_month]"),
+      exp_year: `20${formData.get("order_address[exp_year]")}`,
+    };
+    console.log(card)
```
```
console.log(card)
```
こちらデバック用コードかと存じます。
もう不要であるため削除しておきましょう。

AsukaToyoshima commented on app/javascript/card.js:
```diff
@@ -0,0 +1,31 @@
+    console.log(card)
+    Payjp.createToken(card, (status, response) => {
+      if (status === 200) {
+        debugger
```
```
debugger
```
こちらもデバックコードかと存じます。不要であるため削除しておきましょう。

AsukaToyoshima commented on app/models/order_address.rb:
```diff
@@ -0,0 +1,22 @@
+  include ActiveModel::Model
+  attr_accessor :item_id, :user_id, :post_code, :prefecture_id, :city, :block, :building, :phone_number, :token
+
+  validates :token, presence: true
+  validates :prefecture_id, numericality: { other_than: 1, message: "can't be blank" }
+  validates :post_code, presence: true
+  validates :post_code, format: { with: /\A[0-9]{3}-[0-9]{4}\z/, message: 'is invalid. Enter it as follows (e.g. 123-4567)' }
+  validates :city, presence: true
+  validates :block, presence: true
+  validates :phone_number, presence: true
+  validates :phone_number, format: { with: /\A[0-9]+\z/, message: 'is invalid. Input only number' }
+  validates :phone_number, format: { with: /\A\d{10,11}\z/, message: 'is too short' }
+  validates :user_id, presence: true
+  validates :item_id, presence: true
```
こちら、同じバリデーションを記述しているため、with_optionsを使用して、まとめてみましょう。
下記をご参考ください。
```Ruby
with_options presence: true do
    validates :カラム名
    validates :カラム名
　　・・・
end
```

AsukaToyoshima commented on spec/models/order_address_spec.rb:
```diff
@@ -0,0 +1,90 @@
+        @order_address.phone_number = '123456789'
+        @order_address.valid?
+        expect(@order_address.errors.full_messages).to include('Phone number is too short')
+      end
```
今回の要件、設定したバリデーションに合わせ、以下のテストを追加しましょう。
```
 - 電話番号が12桁以上では購入できない
```

AsukaToyoshima reviewed (commented):
レビューのご依頼ありがとうございます。
コメントさせていただきましたので、ご確認の上、修正をお願いいたします。

疑問点・不明点等ございましたら、コネクトメンターへお気軽にご質問ください。

yamachan32 committed:
Issue10 商品購入機能 レビュー内容の修正

- card.jsから不要なデバックコードの削除
- order_addressモデルのバリエーションをまとめて記述
- 単体テストコードに12桁以上の入力の条件を追加

yamachan32 commented:
@AsukaToyoshima さん

早速のレビュー、ありがとうございました。
デバックコードが残っていた件につきましては、注意不足で申し訳ありませんでした。

２点ほど確認したい件がございますので、後日メンターさんへ確認させていただきたいと思います。

ご指摘いただいた点を修正しましたので、再度レビューをお願いできればと思います。
よろしくお願いします。

AsukaToyoshima commented on app/controllers/items_controller.rb:
```diff
@@ -23,6 +24,7 @@ def show
   end
 
   def edit
+    redirect_to root_path if Order.exists?(item_id: params[:id])
```
【修正任意】
```
if Order.exists?(item_id: params[:id])
```
Ordersテーブルを検索することで、売れているかどうかを判定していただいておりますね。
こちら、アソシエーションを使用することも可能です。
別解としてご参考ください。
```
if @item.order.present?
```

AsukaToyoshima commented on app/controllers/orders_controller.rb:
```diff
@@ -0,0 +1,39 @@
+  end
+
+  def create
+    @payment = Item.find(params[:item_id]).price
```
【修正任意】
```
@payment = Item.find(params[:item_id]).price
```
こちらの記載を`@item = Item.find(params[:item_id])`とし、
34行目で`@item.price`と記述することも可能です。

上記をおこない、
5行目の`@item = Item.find(params[:item_id])`とまとめてメソッド化、before_actionで呼び出すことで、
よりDRYを意識したコードとすることが可能です。

AsukaToyoshima commented on app/models/order_address.rb:
```diff
@@ -0,0 +1,25 @@
+  include ActiveModel::Model
+  attr_accessor :item_id, :user_id, :post_code, :prefecture_id, :city, :block, :building, :phone_number, :token
+
+  with_options presence: true do
+    validates :token
+    validates :post_code
+    validates :city
+    validates :block
+    validates :phone_number
+    validates :user_id
+    validates :item_id
+  end
+
+  validates :prefecture_id, numericality: { other_than: 1, message: "can't be blank" }
+  validates :post_code, format: { with: /\A[0-9]{3}-[0-9]{4}\z/, message: 'is invalid. Enter it as follows (e.g. 123-4567)' }
+  validates :phone_number, format: { with: /\A[0-9]+\z/, message: 'is invalid. Input only number' }
+  validates :phone_number, format: { with: /\A\d{10,11}\z/, message: 'is too short' }
```
【修正任意】
with_optionsを使用してまとめていただきましたね。
下記のように、カラム名の後ろに、そのカラム名のみにつけたいバリデーションを記載することも可能です。
```
  with_options presence: true do
    validates :token
    validates :post_code, format: { ・・・ }
    ・・・
  end
```
上記ですと、post_codeのみにformatのバリデーションを付与することができます。

AsukaToyoshima replied:
【修正任意】
```
  validates :phone_number, format: { with: /\A[0-9]+\z/, message: 'is invalid. Input only number' }
  validates :phone_number, format: { with: /\A\d{10,11}\z/, message: 'is too short' }
```
「電話番号が数値のみであること」「電話番号の桁数が10~11であること」のエラーメッセージを切り替えるために、上記バリデーションをわけていただいたかと存じます。

`/\A\d{10,11}\z/`には「数字のみが使用されていること」といった意味もあるため、`/\A[0-9]+\z/`と重複する内容となってしまいます。

これを解決するには下記2つの方法があります。
・2行目を文字数のみをチェックするバリデーションにする
・1行目を削除し、2行目で数値と文字数を両方チェックしてしまう（エラー文は1種類しか表示させないようにする）
　※今回は仕様としてエラー文の指定がないため、まとめてもOKです

AsukaToyoshima commented on app/views/items/show.html.erb:
```diff
@@ -22,15 +22,15 @@
-        <%# 商品が売れていない場合はこちらを表示しましょう %>
-        <%= link_to "購入画面に進む", "#" ,class:"item-red-btn"%>
-        <%# //商品が売れていない場合はこちらを表示しましょう %>
+    <% unless Order.exists?(item_id: @item.id)%>
+      <% if user_signed_in? %>
+        <% if @item.user_id == current_user.id %>
+          <%= link_to "商品の編集", edit_item_path(@item.id), method: :get, class: "item-red-btn" %>
+          <p class="or-text">or</p>
+          <%= link_to "削除", item_path(@item.id), method: :delete, class:"item-destroy" %>
+        <% else  %>
+          <%= link_to "購入画面に進む", item_orders_path(@item.id) ,class:"item-red-btn"%>
+        <% end %>
```
【修正任意】
下記のような条件文も記載することが可能です。
入れ子を1つ減らすことができ、可読性も向上します。ご参考ください。
```Ruby
if ユーザーがログインしている かつ 商品は購入されていない
　 if ログインユーザー＝商品の出品者
	編集ボタンの表示
	削除ボタンの表示
 else
	購入ボタンの表示
　 end
end
```

AsukaToyoshima reviewed (commented):
レビューのご依頼ありがとうございます。
こちらLGTMとさせていただきます。お疲れ様でした！

コメントをつけさせていただいておりますが、こちらの修正は任意となります。
より良いコードを書くための知識として、ご参考いただけますと幸いです。
修正いただいた場合も、追加でのコードレビューは不要となります。

確認したいことがあるとのことですので、またお時間があるときにでもコネクトメンターへお気軽にご連絡ください。
最終課題完成まであと少し、応援しております！

yamachan32 commented:
@AsukaToyoshima さん

LGTMありがとうございます！
併せて、コードの改善点もご指摘くださりありがとうございます！
ご指摘もとてもわかりやすく、その方が可読性が上がることも理解できました！
そのように修正させていただきたいと思います。
今後とも、どうぞよろしくお願いします🙇‍♂️

yamachan32 committed:
Issue10 商品購入機能 レビュー内容の修正

- Order.exists?を.order.present?に変更
- @item = Item.find(params[:item_id])メソッド化
- order_addressモデルのバリデーションの記述を整理
- order_addressモデルの電話番号正規表現を修正
